pkg/volume/csi/csi_driver_reg.go

```diff
@@ -0,0 +1,288 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/golang/glog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	// Name of node annotation that contains JSON map of driver names to node
+	// names
+	annotationKey = "csi.volume.kubernetes.io/nodeid"
+)
+
+type csiDriver struct {
+	driverName     string
+	driverEndpoint string
+}
+
+// csiDrivers map keep track of all registered CSI drivers on the node and their
+// corresponding sockets
+var csiDrivers sync.Map
+
+// labelManagement is struct of channels used for communication between the driver registration
+// code and the go routine responsible for managing the node's labels
+type labelManagement struct {
+	in  chan csiDriver
+	out chan bool
+}
+
+var lm labelManagement
+
+// Initalizing csiDrivers map and label management channels
+func init() {
```

Comment:
Avoid using package init.  Use a constructor function (see previous)

Reply:
The reason I am using `init` func is because I need to make sure that initialization of sync.Map and communication channels will happen before anything. I do not see any other way to ensure it.

Reply:
Understood, this maybe a good place - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_plugin.go#L60

Reply:
I was under impression that csi_plugin's Init func will be called by multiple components: kubelet, controller_manager etc.. So then I would need to make sure the constructor for  labelManager is not called multiple times. I am not sure how reliable it would be?